src/lib/storage/fsst_segment.cpp

```diff
@@ -0,0 +1,167 @@
+#include "fsst_segment.hpp"
+
+#include <climits>
+#include <sstream>
+#include <string>
+#include "resolve_type.hpp"
+#include "storage/vector_compression/base_compressed_vector.hpp"
+#include "storage/vector_compression/base_vector_decompressor.hpp"
+#include "storage/vector_compression/resolve_compressed_vector_type.hpp"
+#include "utils/assert.hpp"
+#include "utils/performance_warning.hpp"
+
+#include <libfsst.hpp>
+
+namespace opossum {
+
+template <typename T>
+FSSTSegment<T>::FSSTSegment(pmr_vector<unsigned char>& compressed_values,
+                            std::unique_ptr<const BaseCompressedVector>& compressed_offsets,
+                            pmr_vector<uint64_t>& reference_offsets, std::optional<pmr_vector<bool>>& null_values,
+                            uint64_t number_elements_per_reference_bucket, fsst_decoder_t& decoder)
+    : AbstractEncodedSegment{data_type_from_type<pmr_string>()},
+      _compressed_values{std::move(compressed_values)},
+      _compressed_offsets{std::move(compressed_offsets)},
+      _reference_offsets{std::move(reference_offsets)},
+      _null_values{std::move(null_values)},
+      _number_elements_per_reference_bucket{number_elements_per_reference_bucket},
+      _decoder{decoder},
+      _offset_decompressor{_compressed_offsets->create_base_decompressor()} {}
+
+template <typename T>
+AllTypeVariant FSSTSegment<T>::operator[](const ChunkOffset chunk_offset) const {
+  PerformanceWarning("operator[] used");
+  DebugAssert(chunk_offset < size(), "Passed chunk offset must be valid.");
+
+  const auto typed_value = get_typed_value(chunk_offset);
+  if (typed_value) {
+    return typed_value.value();
+  }
+  return NULL_VALUE;
+}
+
+template <typename T>
+uint64_t FSSTSegment<T>::get_offset(const ChunkOffset chunk_offset) const {
+  // For the number of values below 8 (number of reference offsets in encoder) we don't use reference offsets;
+  // In that case _number_elements_per_reference_bucket == 0.
+  // For the first bucket we don't use reference offsets either.
+  if (_number_elements_per_reference_bucket == 0 || chunk_offset < _number_elements_per_reference_bucket) {
+    return _offset_decompressor->get(chunk_offset);
+  }
+
+  // Calculate the corresponding reference offset index for the chunk_offset.
+  auto reference_offset_index = (chunk_offset / _number_elements_per_reference_bucket) - 1;
+  // Merge the last uncompleted bucket with the second last bucket.
```

Comment:
That does not look like a merge. Can you improve the comment and name a case that shows why we need the `min()`?

Reply:
We improved the comment in `fsst_encoder.hpp` and reference it now in `fsst_segment.cpp`.